### PR TITLE
Remove global variables from asmcomp/spill.ml

### PR DESCRIPTION
**Pull request by smuenzel**
Remove the global references/hashtables from spill.ml.
the "inside_loop" ref and the "reset" function in the interface were not in use.

### CODE REVIEW

File: asmcomp/spill.ml [hunk 1]
```diff
@@ -35,35 +35,54 @@ open Mach
...
+(* Record the position of last use of registers *)
+  ; mutable current_date : int
+  ; mutable destroyed_at_fork : (instruction * Reg.Set.t) list
+(* A-list recording what is destroyed at if-then-else points. *)
+  ; mutable spill_at_raise : Reg.Set.t
+  ; mutable inside_arm : bool
+  ; mutable inside_catch : bool
+  ; spill_at_exit : (int, Reg.Set.t) Hashtbl.t
+  ; reload_at_exit : (int, Reg.Set.t) Hashtbl.t
+  }
```
Comment (gasche) [inline asmcomp/spill.ml]: Nitpick: the superior styles for records are the one that gives the cleanest diff for any insertion or reordering of fields. This one, besides being very rare in the codebase, fails for changes touching the first field. (We would need to allow an optional `;` at the beginning to make it suitable). I would use

```
type t = {
  <field>;
  <field>;
}
```

which is commonly used throughout the compiler.

File: asmcomp/spill.ml [hunk 2]
```diff
@@ -35,35 +35,54 @@ open Mach
...
+  { spill_env = Reg.Map.empty
+  ; use_date = Reg.Map.empty
+  ; current_date = 0
+  ; destroyed_at_fork = []
+  ; spill_at_raise = Reg.Set.empty
+  ; inside_arm = false
+  ; inside_catch = false
+  ; spill_at_exit = Hashtbl.create 20
+  ; reload_at_exit = Hashtbl.create 20
+  }
```
Comment (gasche) [inline asmcomp/spill.ml]: (same boring style remark)

**Review by gasche — COMMENTED** [review]
It's not my turf, but I'm sympathetic to the change. The cost in terms of readability is small, the code is closer to how I would naturally write it today I think (no need to carefully reset things). I believe that in general similar changes throughout the backend would make it easier to use as a library / in other contexts.

I think it would be slightly nicer if you had separated the state of the "first pass" (`reload`) and the state of the "second pass" (`spill`). If I read the code right, the only piece of state that is shared between both passes are `spill_env`, and `destroyed_at_fork` which is written in the first pass and read in the second. The types could tell us that:

```ocaml
type reload_data = {
  spill_env : Reg.t Reg.Map.t ref;
  mutable use_date : int Reg.Map.t;
    (* Record the position of last use of registers *)
  mutable current_date : int;
  mutable destroyed_at_fork : (instruction * Reg.Set.t) list;
    (* A-list recording what is destroyed at if-then-else points. *)
  reload_at_exit : (int, Reg.Set.t) Hashtbl.t;
}

type spill_data = {
  spill_env : Reg.t Reg.Map.t ref;
  destroyed_at_fork : (instruction * Reg.Set.t) list;
   (* A-list recording what is destroyed at if-then-else points. *)
  mutable spill_at_raise : Reg.Set.t;
  mutable inside_arm : bool;
  mutable inside_catch : bool;
  spill_at_exit : (int, Reg.Set.t) Hashtbl.t;
}
```

**Comment by smuenzel** [comment]
Separating the types is a good idea, I tried doing that the way you proposed, but there are two things that speak against it:

- `add_spills` is called both from `spill` and `reload`, which means we need to share the data again, or duplicate some code to operate on both spill and reload data
- we need to add type annotations to many functions so that type-based record disambiguation does the right thing.

File: asmcomp/spill.ml (hunk 1 above)
Comment (smuenzel) [reply asmcomp/spill.ml]: +1 for the optional `;`

I made the change to the other style

**Comment by smuenzel** [comment]
Thanks for the review!

> I believe that in general similar changes throughout the backend would make it easier to use as a library / in other contexts.

Definitely agreed. I think setting the scope of `reset` explicitly will also help in files where only some refs should/need to be reset (I believe @hhugo found one of these when looking around the backend for refs).

I don't think this needs an entry in the changelog, since there are no user-visible changes?

**Comment by gasche** [comment]
> `add_spills` is called both from `spill` and `reload`, which means we need to share the data again, or duplicate some code to operate on both spill and reload data

`add_spills` and `spill_reg` take a `spill_env : Reg.t Reg.Map.t ref` as state parameter.

**Comment by smuenzel** [comment]
Ah got it. It feels a little less "clean" to me, but I pushed the change.

**Review by gasche — APPROVED** [review]
Thanks!

I'll wait a few days in case other people have opposite opinions about the refactoring. Otherwise I'll merge by the end of next week if I remember to, or whenever I'm pinged after that.

**Comment by gasche** [comment]
Merged, thanks!

**Pull request merged**